### ISSUE TITLE
Pure BBB integration

### ISSUE THREAD

**Issue by sualko** [ramezrafla/spreed-bigbluebutton#10]
Hi everyone,

I just want to inform you that I'm working on something similar: https://github.com/sualko/cloud_bbb

Didn't find this repo earlier (thx @Ludovicis), but maybe we can cooperate. What do you think? What are your aims? What is the benefit of changing the default nc talk behavior compared to a standalone integration.

**Comment by ramezrafla** [comment]
Hi @sualko,
 
Sorry for delay

I saw the message too about your development on the forum. To be honest, I haven't yet given any thought yet about how to merge our two efforts (do you open up the BBB HTML5 client in a new tab / iframe? Didn't see any info about that).

Our main objective was to have something like NC Talk but with BBB. NC Talk is well integrated in the UI (e.g. file-side chat). We are not going to have both NC Talk and this app, so a single that supported BBB was the way to go for us.

Our short-term goal is to push into the apps store very soon with a new name.
Long-term goal is to add more BBB features in the UI (e.g. recording, uploading a starting presentation, copying the shared etherpad in BBB into the Talk chat).
We are also pulling in the updates from the original NC Talk to keep this app up to date.

How does that match with your objectives?

Is there a way to **merge both repos** into a single **Talk with BBB**

**Comment by sualko** [comment]
> do you open up the BBB HTML5 client in a new tab

I basically replace/integrate the functionality of Greenlight into Nextcloud. This means you can create rooms and share the link e.g. via url files. You can also start a room with a defined presentation from the files app.

> How does that match with your objectives?

My goal was it to create a lightweight BBB integration as soon as possible which is easy to extend. I see the advantage of e.g. not having two chats (one with talk, one in the room), but in my opinion the overhead to modify and merge upstream changes additionally to integrate new features is way to much. Maybe the other way around would be better. Load the original talk into BBB.

Also I could imagine that a lot of people only want to use BBB and don't want to have all the other talk functionality, but that's only a wild guess.

It would be great if we could work together, because I have many other open source projects and only a limited amount of free time.

**Comment by ramezrafla** [comment]
> I basically replace/integrate the functionality of Greenlight into Nextcloud

So you are adding a layer on top of Greenlight. Then the link is shared. In other words an additional manual step for users. Our goal was a seamless integration with BBB, as our clients are not techies, they want to click and talk.

> You can also start a room with a defined presentation from the files app.

This is **GREAT**! I also wanted a way to upload recording back to the files app.

> Load the original talk into BBB

I don't know if that is possible with the BBB API. Maybe the other way around, load the BBB talk into NC Talk.

There are clashes between BBB client and NC talk (the chat is one of them, location of buttons is another). So a proper solution could be a customised BBB client, otherwise chat history is lost.

> It would be great if we could work together, because I have many other open source projects and only a limited amount of free time.

I agree, two similar projects are not in the interest of the community.

**An idea**:
I like your file sharing approach, and this is missing from this fork. But I do like the fact that we can use NC Talk features (integration with project, chat etc.). If someone wanted a standalone BBB then might as well use Greenlight and they are done. 

Can you see a way for us to start with NC Talk as our backbone and add features to it together? I am struggling with my limited knowledge of publishing this app to the apps store as there are namespace clashes.

**Comment by sualko** [comment]
> So you are adding a layer on top of Greenlight

No. I try to add the same functionality to Nextcloud.

> This is GREAT! I also wanted a way to upload recording back to the files app.

This feature is not related to recordings. In my implementation you can choose the presentation instead that you have to upload it inside the BBB client.

> So a proper solution could be a customised BBB client, otherwise chat history is lost.

If I understand you correctly than you have two chats currently in your solution. Than I don't understand why you integrated BBB into talk.

> Can you see a way for us to start with NC Talk as our backbone and add features to it together? 

I don't see any advantage in integrating (to be more precise, replacing the video backend) BBB into talk, but I see a lot of issues in maintaining a fork with a lot of changes. Using all NC interfaces to integrate BBB into projects and so on is probably easier and allows a faster progress, specially new features.

Are there any other opinions from the communicty? @Ludovicis @lars-becker @avosskuehler

**Comment by avosskuehler** [comment]
From my point of view the @ramezrafla approach does fit if you are on a multi user NC environment like a workgroup that is willing to have a BBB conference with users from an existing chat. In my special case of education all classes that are joined together in different talk chats for daily communication can now also start a video conference without having to share links, just look who is online, start a call and talk, even with >30 users. I would prefer to have a check box in the spreed configuration for using the spreed with a bbb server, but it would also be ok to have an extra nc module if the talk updates are merged into. It seems to me that the @sualko approach focuses more on a single user who targets video talks with different audiences. So in my opinion there are less intersections between the two projects on first sight and no urgent need to merge, cause technologically they seem to be very different (but i haven´t got a clue ...)

**Comment by lars-becker** [comment]
> I don't see any advantage in integrating (to be more precise, replacing the video backend) BBB into talk, but I see a lot of issues in maintaining a fork with a lot of changes. Using all NC interfaces to integrate BBB into projects and so on is probably easier and allows a faster progress, specially new features.

I share your point of view. I do some voluntary administration for three non-profit educational organisations. One of them uses Talk for chat and since NC18 workflows for notifications or video while working with an office document in Collabora Office. I can't really see how BBB can replace those integrations in a maintainable way. If it could be done (I only have a very rough overview over the architecture of both Talk and BBB) it seems quite likely that it would need **a lot** of work to implement BBB as a substitute and to maintain it.  Talk has it's merits when used without video or only two or three users. Bottom line: one important reason why I decided to use your approach was that I'm less sure that @ramezrafla approach can be maintained in the long run.

I really like your more lightweight approach. @avosskuehler made one point, which I share: currently the management of rooms on a per user basis is not ideal for organisations where more people prepare / manage the same virtual "events". But this is a minor nitpick and could be addressed with your lightweight approach. One doesn't need to replace Talk for this.

**Comment by ramezrafla** [comment]
@sualko @lars-becker 
I am answering some of the concerns above
1. We are using BBB only for the main talk interface not the side chat in files panel; that functionality is remaining as-is as I also don't see a reason to use BBB here.
2. In terms of maintenance and maturity, it's already production-ready and works (besides us, other members of the community are using it). Maintenance, like any open-source project requires dedication. Pulling in changes from the main repo is not that hard, and we just did it for 8.0.8 the very next day. We are using it in our product so are committed to maintaining in the foreseeable future. If one day decide to stop, you can revert back to the original Talk with no loss of data or conflicts

I agree with @avosskuehler that Talk has some really nice features. Ideally, BBB would be a choice in the original app. This does not seem to be considered at the moment. Right now, we have the best of both worlds:
- A great Talk app with deep integration in NC
- A solid conferencing solution with support for large audiences and a mature UI

@sualko 's approach is great for those only needing a conferencing solution. We need collaboration with conferencing.

**Comment by sualko** [comment]
I think we have two different requirements. You want to replace the costly high performance backend, like e.g. [also other people](https://help.nextcloud.com/t/new-nextcloud-talk-backend-alternative/77348), and I just want to give my users the possibility to create rooms. I still think that it's more promising to integrate BBB via [Talks API](https://nextcloud-talk.readthedocs.io/en/latest/), but both approaches are valid.

> it's already production-ready

I guess this depends on the definition. I just have a few remarks if you like to publish this app in the store:

- make sure that there are no conflicts if people install your app and talk. Especially, because Talk is one of the default apps.
- make sure that this works, in some way, with the Talk mobile apps.
- don't use `putenv`, because it could be blocked on some servers. Extend the bbb base class and set the values in the new constructor (all variables should be protected).

Have fun and thanks to everyone who contributed to this issue.
